### ISSUE TITLE
[explicit-module-boundary-types] False positive on functions nested inside exported

### ISSUE THREAD

**Issue by mijay** [typescript-eslint/typescript-eslint#1486]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": "error"
  }
}
```

```TS
export function outerFunction(): void {
    innerFunction();

    function innerFunction() {
        // do nothing
    }
}
```

**Expected Result**
No linting errors, since `outerFunction` has an explicit type definition.

**Actual Result**
```log
4:5  error  Missing return type on function  @typescript-eslint/explicit-module-boundary-types
```
If you add `: void` to `innerFunction`, the error goes away.

**Additional Info**

It seems to me that the problem is in [this code](https://github.com/typescript-eslint/typescript-eslint/blob/v2.17.0/packages/eslint-plugin/src/rules/explicit-module-boundary-types.ts#L213): it keeps walking up the AST tree till it finds an exported function/class and doesn't stop on local functions, classes, or lambdas.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.17.0` |
| `@typescript-eslint/parser`        | `2.17.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.14.0` |
| `npm`                              | `6.13.6` |
